README.rst
lizard-fancylayers
==========================================

Lizard-fancylayers is an app that tries to be relatively simple (it's
a prototype, not a full fledged app) that visualizes data coming from
lizard-datasource.


Installation notes
------------------

There are a few dependencies that you also need to add to ``INSTALLED_APPS``::

    INSTALLED_APPS = (
        ...
        'lizard_fancylayers',
        'lizard_datasource',
        'colorful',
        ...
        )

In case FEWS JDBC is used as a datasource (the only option as of the
time of writing), there is also a script that runs every minute, to
see if there are new latest values in it. It is run through supervisor.

To the [supervisor] part in Buildout (usually in server.cfg), a line like::

      40 cache_latest_values (autostart=false autorestart=false startsecs=0) ${buildout:bin-directory}/django [cache_latest_values]

should be added. This line makes sure that a command
"bin/supervisorctl start cache_latest_values" is available, which in
turn makes sure that there is always at most one copy of the script
running.

Then, a cronjob that calls that should also be added, and it should
run _every minute_ (almost all the time, the script will immediately
conclude it has nothing to do and exit). E.g.::

    [latest-values-cronjob]
    recipe = z3c.recipe.usercrontab
    times = * * * * *
    command = ${buildout:bin-directory}/supervisorctl start cache_latest_values > /dev/null

Don't forget to add it to parts too, for instance in production.cfg.

Finally, add lizard-fancylayers to the site's urls.py. It's usually
under /fancy/, and that will be used below, but of course that's not
necessary.


Configuration
-------------

As of the time of writing, lizard-fancylayers is usually used to take
some layer defined in FEWS JDBC, and add colors, extra graph lines or
percentiles to it. Here is how to do that.

1. Go to http://yoursitesurl.lizard.net/fancy/

   This searches for available datasources, and as a side effect saves a
   config object that can be edited in the admin for each of them.

2. Create an augmented datasource. The augmented source is the thing
   we want to add things to; if you want to add to the FEWS JDBC source
   with slug "some_slug", choose "'some_slug' from app 'lizard_fewsjdbc'".

3. Go to /admin/lizard_datasource/datasourcemodel/ . Make the
   augmented datasource visible, and all other datasources invisible. Run
   the cache_latest_values script on it (it's an action in the "Action"
   pulldown at the top).

4. Go to /fancy/ again -- you should now see data coming from FEWS, a
   tree of filters to choose from.

5. Go to the datasource layers in the admin, and find all the layers
   you want to do something with. If some combination of
   slug/filter/parameter occurs more than once, pick the one belonging to
   your augmented datasource model. Give the layers you want to use for
   something a nickname.


CHANGES.rst
Changelog of lizard-fancylayers
===================================================


1.1 (unreleased)
----------------

- Nothing changed yet.


1.0 (2014-12-15)
----------------

- Move to mapnik 2.2.0.

- Remove nested 'ui' and 'map' urls when not running standalone.

- Re-added 'add to collage/dashboard' link.

- Marked the graph as zoomable: this enables lizard-fewsjdbc's douglas-peucker
  behaviour.

- Add the function values_multiple_timeseries() to get all timeseries.

- Omitting double or triple results if they're the same.

- Translation fixes.  
  

0.10 (2013-06-06)
-----------------

- Use new lizard-datasource method to show labels and y-axes.


0.9 (2013-04-23)
----------------

- Fix bug with configured default icons -- Mapnik can't handle Unicode
  filenames.


0.8 (2013-04-11)
----------------

- Let the standard color and standard icon be configured using
 lizard-map's Settings (FANCYLAYERS_DEFAULT_COLOR,
 FANCYLAYERS_DEFAULT_SYMBOL_NAME and FANCYLAYERS_DEFAULT_SYMBOL_MASK)

- Show first graph in normal color, next graph lines in green

- Work with latest datasource timeseries API

- Get graph legend correct, from timeseries' column names


0.7 (2013-04-10)
----------------

- Make it possible to show several timeseries from one location in one
  graph. Doesn't look nice yet.


0.6 (2013-03-19)
----------------

- Hiding the legend if there's no proper legend to be found.

- Fixed icon before popup graph title.


0.5 (2013-02-12)
----------------

- Implement a legend() function using lizard-datasource's
  location_annotations.


0.4 (2013-01-16)
----------------

- Added a values() method to the adapter, so that CSV files with
  values can be downloaded.

- Show a unit in the popup.

0.3 (2013-01-03)
----------------

- If a popup is shown for several identifiers at once, they are now
  shown as separate graphs, instead of all timeseries in the same
  graph. Mainly because the combined graphs got very messy once
  percentiles were involved.


0.2 (2012-12-20)
----------------

- Add support for percentiles, colors.


0.1 (2012-11-15)
----------------

Not really working yet, but making a release for test purposes.


lizard_fancylayers/tests.py
# (c) Nelen & Schuurmans.  GPL licensed, see LICENSE.rst.

from django.test import TestCase




lizard_fancylayers/tests/test_views.py
import mock

from django.contrib.sessions.middleware import SessionMiddleware
from django.contrib.auth.models import AnonymousUser

from django.test import TestCase
from django.test import RequestFactory

from lizard_fancylayers import views


class TestHomepageView(TestCase):
    def setUp(self):
        self.view_function = views.HomepageView.as_view()

    def test_has_response(self):
        request = RequestFactory().get('/')
        middleware = SessionMiddleware()
        middleware.process_request(request)
        request.user = AnonymousUser()

        response = self.view_function(request, '')

        self.assertContains(response, 'Apps overview', status_code=200)


lizard_fancylayers/tests/test_layers.py
import mock

from django.test import TestCase
from lizard_datasource import datasource
from lizard_datasource import location

from lizard_fancylayers import layers


class MockDataSource(datasource.DataSource):
    def locations(self, bare=True):
        return [location.Location('dummy', 0, 0)]

    def unit(self, choices_made=None):
        return None


@mock.patch('lizard_datasource.datasource.get_datasources',
            return_value=[MockDataSource()])
class TestAdapter(TestCase):
    def test_constructor_trivial(self, patched_datasource):
        workspace_item = mock.MagicMock()

        adapter = layers.FancyLayersAdapter(
            workspace_item,
            layer_arguments={'choices_made': "{}"})

        self.assertFalse(adapter.datasource is None)

    def test_html(self, patched_datasource):
        workspace_item = mock.MagicMock()

        adapter = layers.FancyLayersAdapter(
            workspace_item,
            layer_arguments={'choices_made': "{}"})

        html = adapter.html(identifiers=[{'identifier': 'dummy'}])

        self.assertTrue(html)

    def test_returns_legend(self, patched_datasource):
        workspace_item = mock.MagicMock()

        adapter = layers.FancyLayersAdapter(
            workspace_item,
            layer_arguments={'choices_made': "{}"})

        legend = adapter.legend()

        # Must be an iterable
        l = list(legend)

        # Must be something in it
        self.assertTrue(len(l) > 0)

        # Must have right keys
        self.assertTrue('img_url' in l[0])
        self.assertTrue('description' in l[0])

    def test_empty_legend(self, patched_datasource):
        # Don't return a legend if there are no annotations to grab colors
        # from.
        workspace_item = mock.MagicMock()

        adapter = layers.FancyLayersAdapter(
            workspace_item,
            layer_arguments={'choices_made': "{}"})

        legend = adapter.legend()
        self.assertEquals(legend, [])


lizard_fancylayers/urls.py
# (c) Nelen & Schuurmans.  GPL licensed, see LICENSE.rst.
from django.conf.urls import include
from django.conf.urls import patterns
from django.conf.urls import url
from django.contrib import admin
from django.views.i18n import javascript_catalog

from django.conf import settings

from lizard_ui.urls import debugmode_urlpatterns

from lizard_fancylayers import views

admin.autodiscover()

urlpatterns = patterns(
    '',

    url(r'^(([a-z0-9_]+-[a-zA-Z0-9-._]+/)*)$',
        views.HomepageView.as_view(),
        name="lizard_fancylayers.homepage",
        ),
    )

if getattr(settings, 'FANCYLAYERS_STANDALONE', False):
    urlpatterns += (
        url(r'^ui/', include('lizard_ui.urls')),
        url(r'^map/', include('lizard_map.urls')),
        url(r'^jsi18n/$', javascript_catalog, {
            'packages': ('lizard_fancylayers',)
            }),
    )
    urlpatterns += debugmode_urlpatterns()


lizard_fancylayers/views.py
# (c) Nelen & Schuurmans.  GPL licensed, see LICENSE.rst.
from __future__ import unicode_literals

import json
import logging

from django.core.urlresolvers import reverse

from lizard_map.lizard_widgets import WorkspaceAcceptable
from lizard_map.views import AppView
from lizard_datasource import datasource

logger = logging.getLogger(__name__)


class FancyLayersView(AppView):
    # Placeholder for our own utils later
    pass


class HomepageView(FancyLayersView):
    template_name = 'lizard_fancylayers/homepage.html'

    def make_url(self, choices_made):
        options = "".join("{0}-{1}/".format(
            item[0], item[1]) for item in choices_made.items())
        return reverse("lizard_fancylayers.homepage", args=(options,))

    def dispatch(self, request, choices, *args):
        self.choices_made = self.make_choices_made(choices)
        return super(HomepageView, self).dispatch(request)

    def make_choices_made(self, choices):
        choicelist = choices.split('/')[:-1]

        choicedict = {}
        for choice in choicelist:
            identifier, value = choice.split('-')
            choicedict[identifier] = value

        return datasource.ChoicesMade(dict=choicedict)

    @property
    def datasource(self):
        if not hasattr(self, '_datasource'):
            self._datasource = datasource.datasource(self.choices_made)
        return self._datasource

    def visible_criteria(self):
        try:
            datasource = self.datasource
            if datasource is None:
                return []
            criteria = datasource.visible_criteria()
            for crit in criteria:
                criterion = crit['criterion']
                options = crit['options']

                for option in options.iter_options():
                    # Suppose we chose this value
                    new_choices_made = self.choices_made.add(
                        criterion.identifier,
                        option.identifier)
                    # Could we draw it on the map then?
                    if datasource.is_drawable(new_choices_made):
                        # Make it a workspace-acceptable
                        option.workspace_acceptable = WorkspaceAcceptable(
                            name=option.description,
                            adapter_name='adapter_fancylayers',
                            adapter_layer_json=json.dumps({
                                'choices_made': new_choices_made.json()}))
                    else:
                        # Give it a URL that makes it choosable
                        option.url = self.make_url(new_choices_made)

            return criteria
        except Exception, e:
            logger.warn("Caught exception in visible_criteria: {0}".
                        format(e))

    def forgettable_criteria(self):
        forgettable_criteria = []
        datasource = self.datasource
        if datasource is None:
            return []
        for criterion in self.datasource.criteria():
            if criterion.identifier not in self.choices_made:
                continue
            forgettable_criteria.append({
                    'identifier': criterion.identifier,
                    'description': criterion.description,
                    'url': self.make_url(
                        self.choices_made.forget(criterion.identifier)),
                    'value': self.choices_made[criterion.identifier]
                    })

        return forgettable_criteria


lizard_fancylayers/testsettings.py
import os

from lizard_ui.settingshelper import setup_logging
from lizard_ui.settingshelper import STATICFILES_FINDERS

DEBUG = True
TEMPLATE_DEBUG = True

# SETTINGS_DIR allows media paths and so to be relative to this settings file
# instead of hardcoded to c:\only\on\my\computer.
SETTINGS_DIR = os.path.dirname(os.path.realpath(__file__))

# BUILDOUT_DIR is for access to the "surrounding" buildout, for instance for
# BUILDOUT_DIR/var/static files to give django-staticfiles a proper place
# to place all collected static files.
BUILDOUT_DIR = os.path.abspath(os.path.join(SETTINGS_DIR, '..'))
LOGGING = setup_logging(BUILDOUT_DIR)

# ENGINE: 'postgresql_psycopg2', 'postgresql', 'mysql', 'sqlite3' or 'oracle'.
# In case of geodatabase, prepend with:
# django.contrib.gis.db.backends.(postgis)
DATABASES = {
    # If you want to use another database, consider putting the database
    # settings in localsettings.py. Otherwise, if you change the settings in
    # the current file and commit them to the repository, other developers will
    # also use these settings whether they have that database or not.
    # One of those other developers is Jenkins, our continuous integration
    # solution. Jenkins can only run the tests of the current application when
    # the specified database exists. When the tests cannot run, Jenkins sees
    # that as an error.
    'default': {
        'NAME': os.path.join(BUILDOUT_DIR, 'var', 'sqlite', 'test.db'),
        'ENGINE': 'django.contrib.gis.db.backends.spatialite',
        'USER': '',
        'PASSWORD': '',
        'HOST': '',  # empty string for localhost.
        'PORT': '',  # empty string for default.
        }
    }
SITE_ID = 1
SECRET_KEY = 'This is not secret but that is ok.'
INSTALLED_APPS = [
    'lizard_fancylayers',
    'lizard_fewsjdbc',
    'lizard_datasource',
    'lizard_ui',
    'lizard_map',
    'lizard_maptree',
    'lizard_wms',
    'django.contrib.staticfiles',
    'compressor',
    'south',
    'django_nose',
    'lizard_security',
    'django_extensions',
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.gis',
    'django.contrib.sites',
    ]
ROOT_URLCONF = 'lizard_fancylayers.urls'

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'lizard_security.middleware.SecurityMiddleware',
    'tls.TLSRequestMiddleware',
    )

TEST_RUNNER = 'django_nose.NoseTestSuiteRunner'

# Used for django-staticfiles (and for media files
STATIC_URL = '/static_media/'
MEDIA_URL = '/media/'
STATIC_ROOT = os.path.join(BUILDOUT_DIR, 'var', 'static')
MEDIA_ROOT = os.path.join(BUILDOUT_DIR, 'var', 'media')
STATICFILES_FINDERS = STATICFILES_FINDERS

FANCYLAYERS_STANDALONE = True

SECRET_KEY = "Does not need to be secret"

try:
    # Import local settings that aren't stored in svn/git.
    from lizard_fancylayers.local_testsettings import *
except ImportError:
    pass


lizard_fancylayers/layers.py
import datetime
import logging
import mapnik
import math
import os
import pytz

from django.conf import settings
from django.http import Http404
from django.template.loader import render_to_string

from lizard_map import coordinates
from lizard_map import workspace
from lizard_map.models import Setting

from lizard_map.adapter import Graph, FlotGraph
from lizard_map.adapter import adapter_serialize
from lizard_map.mapnik_helper import add_datasource_point
from lizard_map.models import ICON_ORIGINALS
from lizard_map.symbol_manager import SymbolManager

from lizard_datasource import properties
from lizard_datasource import datasource
from lizard_datasource import functools

logger = logging.getLogger(__name__)

DEFAULT_COLOR = '0000ff'
DEFAULT_SYMBOL_NAME = 'meetpuntPeil.png'
DEFAULT_SYMBOL_MASK = 'meetpuntPeil_mask.png'

# Used for graph lines after the first
COLORS = (
    'rgb(126, 209, 37)',  # Greenish, and needed for a project
    'rgb(255, 0, 0)',
    'rgb(0, 0, 255)',
    'rgb(0, 255, 255)',
    'rgb(255, 0, 255)',
    'rgb(255, 255, 0)'
)


def default_color():
    return Setting.get("FANCYLAYERS_DEFAULT_COLOR") or DEFAULT_COLOR


def default_symbol_name():
    return (Setting.get("FANCYLAYERS_DEFAULT_SYMBOL_NAME")
            or DEFAULT_SYMBOL_NAME).encode('utf8')


def default_symbol_mask():
    return (Setting.get("FANCYLAYERS_DEFAULT_SYMBOL_MASK")
            or DEFAULT_SYMBOL_MASK).encode('utf8')


@functools.memoize
def html_to_mapnik(color):
    if color[0] == '#':
        color = color[1:]

    r, g, b = color[0:2], color[2:4], color[4:6]
    rr, gg, bb = int(r, 16), int(g, 16), int(b, 16)

    return rr / 255.0, gg / 255.0, bb / 255.0, 1.0


def symbol_filename(color):
    symbol_manager = SymbolManager(
        ICON_ORIGINALS,
        os.path.join(settings.MEDIA_ROOT, 'generated_icons'))
    filename = symbol_manager.get_symbol_transformed(
        default_symbol_name(), mask=(default_symbol_mask(),),
        color=color)
    return filename


def symbol_pathname(color):
    filename = symbol_filename(color)
    pathname = os.path.join(
        settings.MEDIA_ROOT, 'generated_icons', filename)
    return pathname


def symbol_url(color):
    return "{media_url}generated_icons/{filename}".format(
        media_url=settings.MEDIA_URL, filename=symbol_filename(color))


class FancyLayersAdapter(workspace.WorkspaceItemAdapter):
    """Registered as adapter_fancylayers."""

    def __init__(self, *args, **kwargs):
        super(FancyLayersAdapter, self).__init__(*args, **kwargs)

        self.choices_made = datasource.ChoicesMade(
            json=self.layer_arguments['choices_made'])
        self.datasource = datasource.datasource(
            choices_made=self.choices_made)

    def layer(self, layer_ids=None, webcolor=None, request=None):
        # We only do point layers right now
        if not self.datasource.has_property(properties.LAYER_POINTS):
            return [], {}

        layers = []
        styles = {}

        locations = list(self.datasource.locations())
        colors = {"default": html_to_mapnik(default_color())}

        for location in locations:
            if location.color is not None:
                colors[location.color] = html_to_mapnik(location.color)

        style = mapnik.Style()

        for colorname, color in colors.iteritems():
            rule = mapnik.Rule()
            symbol = mapnik.PointSymbolizer()
            symbol.filename = symbol_pathname(color)
            symbol.allow_overlap = True
            rule.symbols.append(symbol)
            rule.filter = mapnik.Filter("[Color] = '{0}'".format(colorname))
            style.rules.append(rule)

        styles['trivialStyle'] = style

        layer = mapnik.Layer("Fancy Layers layer", coordinates.WGS84)
        layer.datasource = mapnik.MemoryDatasource()
        layer.styles.append('trivialStyle')

        for _id, location in enumerate(locations):
            color = location.color or 'default'
            add_datasource_point(
                layer.datasource, location.longitude,
                location.latitude,
                'Color', str(color), _id)

        layers.append(layer)
        return layers, styles

    def search(self, google_x, google_y, radius=None):
        """Return list of dict {'distance': <float>, 'timeserie':
        <timeserie>} of closest fews point that matches x, y, radius.
        """
        def distance(x1, y1, x2, y2):
            return math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)

        locations = self.datasource.locations()
        seen_identifiers = []
        result = []
        for location in locations:
            identifier = location.identifier
            if identifier in seen_identifiers:
                # We constantly get three identical results.
                continue
            seen_identifiers.append(identifier)

            x, y = coordinates.wgs84_to_google(
                location.longitude,
                location.latitude)
            dist = distance(google_x, google_y, x, y)

            if dist < radius:
                result.append(
                    {'distance': dist,
                     'name': location.description(),
                     'shortname': location.identifier,
                     'workspace_item': self.workspace_item,
                     'identifier': {'identifier': identifier},
                     'google_coords': (x, y),
                     'object': None})
        result.sort(key=lambda item: item['distance'])
        return result[:3]  # Max 3.

    def html(self, identifiers=None, layout_options=None):
        """Adapted version of lizard-map's html_default. If there are
        several graphs to be shown, we show them as several graphs in
        one popup tab."""
        is_collage = layout_options and layout_options.get('is_collage', False)

        # Build "adapter-image" url for current adapter and identifiers,
        # one for each identifier.
        urls = [
            {'image_url': self.workspace_mixin_item.url(
                    "lizard_map_adapter_image", [identifier]),
             'flot_url': self.workspace_mixin_item.url(
                    "lizard_map_adapter_flot_graph_data", [identifier]),
             'title': self.location(**identifier)['name']
             }
            for identifier in identifiers]

        # Makes it possible to create collage items from current
        # selected objects.
        collage_item_props = []
        # No export and selection for collages.
        if not is_collage:
            for identifier in identifiers:
                location = self.location(**identifier)
                collage_item_props.append(
                    {'name': location['name'],
                     'adapter_class': self.workspace_mixin_item.adapter_class,
                     'adapter_layer_json':
                         self.workspace_mixin_item.adapter_layer_json,
                     'identifier': adapter_serialize(identifier),
                     'url': self.workspace_mixin_item.url(
                            "lizard_map_adapter_values", [identifier, ],
                            extra_kwargs={'output_type': 'csv'})})

        render_kwargs = {
            'unit': u'',  # Don't show unit above the graphs
            'urls': urls,
            'symbol_url': self.symbol_url(),
            'collage_item_props': collage_item_props,
            'adapter': self,
            }

        if layout_options is not None:
            render_kwargs.update(layout_options)

        return render_to_string(
            'lizard_fancylayers/popup.html',
            render_kwargs)

    def symbol_url(self):
        # Note: it calls the symbol_url *function* in this module.
        return symbol_url(html_to_mapnik(DEFAULT_COLOR))

    def location(self, identifier, layout=None):
        locations = self.datasource.locations()
        for location in locations:
            if location.identifier == identifier:
                break
        else:
            return None

        google_x, google_y = coordinates.wgs84_to_google(
            location.longitude, location.latitude)

        identifier_to_return = {
            'identifier': identifier
            }
        if layout is not None:
            identifier_to_return['layout'] = layout

        description = location.description()

        return {
            'google_coords': (google_x, google_y),
            'name': description,
            'shortname': description,
            'workspace_item': self.workspace_item,
            'identifier': identifier_to_return,
            'object': location
            }

    def values_multiple_timeseries(self, identifier, start_date, end_date):
        """Returns a lizard_datasource.timeseries.Timeseries object."""
        return self.datasource.timeseries(
            identifier['identifier'], start_date, end_date)

    def values(self, identifier, start_date, end_date):
        """Return values in list of dictionaries (datetime, value, unit)
        """
        timeseries = self.datasource.timeseries(
            identifier['identifier'], start_date, end_date)

        values = [{
                'datetime': data[0],
                'value': data[1],
                'unit': ""
                }
                for data in timeseries.data()
                ]

        return values

    def image(
        self, identifiers, start_date, end_date, width=380.0, height=250.0,
        layout_extra=None, raise_404_if_empty=False):
        # Initial version taken from lizard-fewsjdbc

        return self._render_graph(
            identifiers, start_date, end_date, width=width, height=height,
            layout_extra=layout_extra, raise_404_if_empty=raise_404_if_empty,
            GraphClass=Graph)

    def flot_graph_data(
        self, identifiers, start_date, end_date, layout_extra=None,
        raise_404_if_empty=False
    ):
        return self._render_graph(
            identifiers, start_date, end_date, layout_extra=layout_extra,
            raise_404_if_empty=raise_404_if_empty,
            GraphClass=FlotGraph)

    def _render_graph(
        self, identifiers, start_date, end_date, layout_extra=None,
        raise_404_if_empty=False, GraphClass=Graph, **extra_params
    ):
        """
        Visualize timeseries in a graph.

        Legend is always drawn.

        New: this is now a more generalized version of image(), to
        support FlotGraph.
        """

        def apply_lines(identifier, values, location_name):
            """Adds lines that are defined in layout. Uses function
            variable graph, line_styles.

            Inspired by fewsunblobbed"""

            layout = identifier['layout']

            if "line_min" in layout:
                graph.axes.axhline(
                    min(values),
                    color=line_styles[str(identifier)]['color'],
                    lw=line_styles[str(identifier)]['min_linewidth'],
                    ls=line_styles[str(identifier)]['min_linestyle'],
                    label='Minimum %s' % location_name)
            if "line_max" in layout:
                graph.axes.axhline(
                    max(values),
                    color=line_styles[str(identifier)]['color'],
                    lw=line_styles[str(identifier)]['max_linewidth'],
                    ls=line_styles[str(identifier)]['max_linestyle'],
                    label='Maximum %s' % location_name)
            if "line_avg" in layout and values:
                average = sum(values) / len(values)
                graph.axes.axhline(
                    average,
                    color=line_styles[str(identifier)]['color'],
                    lw=line_styles[str(identifier)]['avg_linewidth'],
                    ls=line_styles[str(identifier)]['avg_linestyle'],
                    label='Gemiddelde %s' % location_name)

        line_styles = self.line_styles(identifiers)

        locations = list(self.datasource.locations())
        today = datetime.datetime.now()
        graph = GraphClass(
            start_date, end_date, today=today,
            tz=pytz.timezone(settings.TIME_ZONE), **extra_params)
        graph.axes.grid(True)
#        parameter_name, unit = self.jdbc_source.get_name_and_unit(
#            self.parameterkey)
#        graph.axes.set_ylabel(unit)

        # Draw extra's (from fewsunblobbed)
        title = None
        y_min, y_max = None, None

        is_empty = True
        for identifier in identifiers:
            location_id = identifier['identifier']

            location_name = [
                location.description() for location in locations
                if location.identifier == location_id][0]

            timeseries = self.datasource.timeseries(
                location_id, start_date, end_date)

            if timeseries is not None:
                is_empty = False
                # Plot data if available.

                has_percentiles = False
                if (self.datasource.has_percentiles() and
                    hasattr(graph, 'add_percentiles')):
                    has_percentiles = True

                for series_num, series_name in enumerate(timeseries.columns):
                    series = timeseries.get_series(series_name)
                    dates = series.keys()
                    values = list(series)

                    # Hack -- show first line in normal color, every
                    # next line in green.  Because for the first
                    # practical use of this code, we have only one
                    # other line, and it must be green.
                    if series_num == 0:
                        color = line_styles[str(identifier)]['color']
                    else:
                        color = COLORS[(series_num - 1) % len(COLORS)]
                    if values:
                        # Percentiles have only one timeserie which
                        # needs the location label to add the percentiles.
                        label = timeseries.label(series_name)
                        if has_percentiles:
                            label = location_name

                        graph.axes.plot(
                            dates, values,
                            lw=1,
                            color=color,
                            label=label)

                # For the y-label, we take all the units that are not None,
                # each of them once only, and join them with ','
                units = []
                for column in timeseries.columns:
                    unit = timeseries.unit(column)
                    if unit is not None and unit not in units:
                        units.append(unit)
                if units:
                    graph.axes.set_ylabel(', '.join(units))

                if has_percentiles:
                    percentiles = self.datasource.percentiles(
                        location_id, start_date, end_date)
                    opacities = ()
                    if len(percentiles) == 2:
                        opacities = (0.4,)
                    elif len(percentiles) == 4:
                        opacities = (0.4, 0.2)

                    if opacities:
                        graph.add_percentiles(
                            location_name, percentiles, opacities)

            # Apply custom layout parameters.
            if 'layout' in identifier:
                layout = identifier['layout']
                if "y_label" in layout:
                    graph.axes.set_ylabel(layout['y_label'])
                if "x_label" in layout:
                    graph.set_xlabel(layout['x_label'])
                apply_lines(identifier, values, location_name)

        if is_empty and raise_404_if_empty:
            raise Http404

        # Originally legend was only turned on if layout.get('legend')
        # was true. However, as far as I can see there is no way for
        # that to become set anymore. Since a legend should always be
        # drawn, we simply put the following:
        graph.legend()

        # If there is data, don't draw a frame around the legend
        if graph.axes.legend_ is not None:
            graph.axes.legend_.draw_frame(False)
        else:
            # TODO: If there isn't, draw a message. Give a hint that
            # using another time period might help.
            pass

        # Extra layout parameters. From lizard-fewsunblobbed.
        y_min_manual = y_min is not None
        y_max_manual = y_max is not None
        if y_min is None:
            y_min, dummy = graph.axes.get_ylim()
        if y_max is None:
            dummy, y_max = graph.axes.get_ylim()

        if title:
            graph.suptitle(title)

        graph.set_ylim(y_min, y_max, y_min_manual, y_max_manual)

        # Copied from lizard-fewsunblobbed.
        if "horizontal_lines" in layout_extra:
            for horizontal_line in layout_extra['horizontal_lines']:
                graph.axes.axhline(
                    horizontal_line['value'],
                    ls=horizontal_line['style']['linestyle'],
                    color=horizontal_line['style']['color'],
                    lw=horizontal_line['style']['linewidth'],
                    label=horizontal_line['name'])

        graph.add_today()
        return graph.render()

    def legend(self, updates=None):
        """Returns a list of {'img_url':..., 'description':...}
        dictionaries that are used to make a legend for this
        datasource."""

        # If there is a legend, it is in the datasource's location annotations
        annotations = self.datasource.location_annotations() or {}

        if 'color' in annotations and annotations['color']:
            # Colorlegend is a list of (color, description) tuples
            colorlegend = annotations['color']
            return [{
                    'img_url': symbol_url(html_to_mapnik(color)),
                    'description': description}
                    for color, description in colorlegend]

        # If we cannot find any proper colored legend, return nothing.
        # This effectively hides the legend.
        return []
